### ISSUE TITLE
Consider a FormatterConfiguration

### ISSUE THREAD

**Issue by Mike-EEE** [serilog/serilog#723]
I encountered this problem while trying to format a `MethodBase` object in my entries.  It would be nice to be able to provide custom configurations for formatting objects.  That is, whenever an object of a certain type is encountered, Serilog looks to a formatting configuration of some sort to see how to format the object.

So something like:

```
var logger = new LoggerConfiguration()
             .Format
             .With<MethodBase>( m => $"{m.DeclaringType.Name}.{m.Name}" )
             .CreateLogger();
```

Then in the [ScalarValue rendering](https://github.com/serilog/serilog/blob/dev/src/Serilog/Events/ScalarValue.cs#L74), it checks to see if a registration like the above has been made.  If it has, then use it, otherwise, continue with existing functionality.

Thank you for any consideration!

**Comment by nblumhardt** [comment]
Hi Mike,

This is essentially `IScalarConversionPolicy`, which exists in Serilog today but isn't exposed fully via the API. It'd be great to check out what is involved in providing some configuration for it.

 * https://github.com/serilog/serilog/blob/dev/src/Serilog/Core/IScalarConversionPolicy.cs
 * https://github.com/serilog/serilog/blob/dev/src/Serilog/Parameters/PropertyValueConverter.cs#L114

(There are some perf tradeoffs to examine if more than a few of these are expected to be plugged in per app, the current impl is O(n) which is net positive when n <= 7 vs a dictionary, but getting a dictionary in for larger numbers of items will need some clever synchronization to avoid any additional locks/fences...)

Any investigation/proposals would be great! :-)

cc @adamchester

**Comment by Mike-EEE** [comment]
Excellent @nblumhardt!  I will check that out and update this thread on any findings/successes/questions.  Thanks for the pointer!

**Comment by Mike-EEE** [comment]
OK I see what you are saying now.  It is indeed available, but not exposed.  Essentially it would involve adding a new parameter to the [`PropertyValueConverter` Constructor`](https://github.com/serilog/serilog/blob/dev/src/Serilog/Parameters/PropertyValueConverter.cs#L50) and passing it in when the `LoggerConfiguration` creates the `ILogger`.  Since it will involve some work, I still gravitate towards the addition of a new configuration object (`Format` or `Formatting`) and add some awesome magic there (that could ultimately tie into creating `IScalarConversionPolicy` implementations).

**Comment by nblumhardt** [comment]
Hi Mike, this one is still causing me a bit of consternation. Sitting with @merbla  here discussing this, and thinking back to the original reason that `IScalarConversionPolicy` is not exposed...

The premise is that _sinks are responsible for formatting scalars_ is a pretty strong design principle in Serilog. 

E.g., if you have a `TimeSpan`, then the MSSQL sink should be free to format that as a SQL timespan, and a JSON-based sink might format it as ISO.

`Destructure.AsScalar()` exists to let more types through this pipeline, but the underlying mechanism (`IScalarConversionPolicy` etc.) probably shouldn't also take the role of providing string formatting.

What do you think? Given this whole concept any more consideration? Cheers!

**Comment by adamchester** [comment]
Why are reflection types considered to be Scalar?

**Comment by adamchester** [comment]
Just for the sake of history, this is the [PoC for transforming reflection types](https://github.com/serilog/serilog/commit/1e14bb41c5753add0c864b99ab8b473f787737ed) that I made while investigating this. I didn't attempt to turn it into a PR because my testing revealed that allowing reflection types to be destructured results in an `OutOfMemoryException`

**Comment by nblumhardt** [comment]
Aaah right, thanks for the link, that makes the issue a bit clearer (I'd forgotten where the handler for this was plugged in).

Regarding why they're scalar, it was too easy to create explosively large object graphs during destructuring otherwise. Perhaps an `IDestructuringPolicy` would have been a better implementation option - `Type` could come through as a string, and `MemberInfo` would be a structure with `Name` and `DeclaringType` (suspect that was one of the earlier suggestions around this thread). `ParameterInfo` could get some similar special treatment.

Am I on track now? :-)

**Comment by nblumhardt** [comment]
Raised a PR to internalize this interface until we're comfortable with a design (if one is required at all).

**Comment by Mike-EEE** [comment]
I'm really glad this has led to some great discussion!  From my take (as a newb but avid fan of Serilog :smile: ) my whole take is that there should be a mechanism whereby if configured I can always expect certain objects to look a certain way.  I appreciate delegating formatting to the sinks.  However, if I have a certain object, I would expect that object to always be formatted a certain way -- either at the sink level and/or before then anytime that object is rendered, or *both*. :)

If the sinks are formatting, they should also respect the wishes of the implementing application to also format it a certain way before they have their say in it, IMO!

**Comment by Mike-EEE** [comment]
Another thought, FWIW, when I first thought about this feature, I was thinking along the lines of WPF `DataTemplate`s, where you could basically create a template a generic template and no matter where in the application it was viewed, you could always expect it to look a certain way -- if that helps!

**Comment by adamchester** [comment]
Maybe it's me that caused it, but there seems to be a few related issues going on here. I'm going to see if I can write down my understanding of them, and how they're related.

## The example given in this issue
@Mike-EEE created this issue with a suggestion to consider a `FormatterConfiguration` as a potential solution to the way `MethodBase` types are treated and ultimately rendered by most (all?) sinks. I say that because the only example given by @Mike-EEE involves `MethodBase` which is a *reflection type*. Maybe @Mike-EEE has different scenarios where formatting would be necessary, but this is only one we have so far.

Reflection types are: `System.Type`, `MethodBase`, `ParameterInfo`, etc. However, in Serilog, they are defined by [`ReflectionTypesScalarConversionPolicy`](https://github.com/serilog/serilog/blob/32127146a394353eb3a26583b5a2ec5b0c404f25/src/Serilog/Policies/ReflectionTypesScalarConversionPolicy.cs) as derived from either `System.Type` OR `System.Reflection.MemberInfo`.

## What is a Scalar? 

[Serilog wiki says](https://github.com/serilog/serilog/wiki/Structured-Data):
> Scalars are .NET types that can be represented as atomic values; most value types like int fit this description, but so do some reference types like Uri and string
> ...
> Out of the box, Serilog recognises the following list as basic scalar types, regardless of any other policies applied:
> Booleans - bool
> Numerics - byte, short, ushort, int, uint, long, ulong, float, double, decimal
> Strings - string, byte[]
> Temporals - DateTime, DateTimeOffset, TimeSpan
> Others - Guid, Uri
> Nullables - nullable versions of any of the types above

I note the absence of `Type` and `MemberInfo`. Most of these above make sense to me as _atomic_, except for perhaps `Uri` which can be broken apart further in useful ways. I can break apart `Uri` into pieces like `HostName` and `Query` (which I've often wanted to do - so that I can query and filter the logs more easily later). I can think of quite a few extremely useful ways to break apart `MethodBase` or even `Type` to make them more useful. Apparently, @Mike-EEE and @nblumhardt can think of various ways to break them apart too ;)

I think they should not be treated as `Scalar`.

# Now what?

I'm not sure yet. I feel like `Formatting` is not the right solution reflection types. As [@nblumhardt says](https://github.com/serilog/serilog/issues/723#issuecomment-215623341):

> sinks are responsible for formatting scalars is a pretty strong design principle in Serilog

Agreed. Makes sense.

I have a question now: what if *all* types could be transformed early in the Serilog pipeline?

**Comment by Mike-EEE** [comment]
Ah yes @adamchester, I did indeed start out with `MethodBase`, but my intention/direction/ask here is for `any` object, whether it is scalar or not.  That is, a specified/configurable mechanism that is application-specific that allows an implementer to define the "look and feel" of any given object in the .NET universe.  Again I come back to the WPF DataTemplate idiom.  That is very much the same spirit here.

> I have a question now: what if all types could be transformed early in the Serilog pipeline?

This is basically what I am after, yes. :)  Essentially, it should be (in order of priority):
1) Application-specific/defined formatting
2) Sink-specific/defined formatting.

If I as an application developer format a `MethodBase` into a string early in this pipeline, then this ultimately gets rendered as a scalar at the sink level.  Because I am deliberately/consciously setting this format/configuration, then I have explicitly given my "OK" on this override.  Hope this makes sense!

**Comment by adamchester** [comment]
Thanks @Mike-EEE.

Do you think something like this makes sense?

```csharp
Log.Logger = new LoggerConfiguration()
    .Transform.ByType<Uri>(
        transformation: uri => new { Canonical = uri.ToString(), uri.Host, uri.Query })
    .Transform.ByType<Type>(
        predicate: typeof(Type).IsAssignableFrom,
        transformation: t => new { Name = t.Name, t.Namespace, })
    .Transform.ByType<MethodBase>(
        predicate: typeof(MethodBase).IsAssignableFrom,
        transformation: mi => new { DisplaySig = mi.ToString(), mi.ReflectedType, mi.DeclaringType })
    .WriteTo.Sink(new LinqpadDumpSink())
    .CreateLogger();
```

**Comment by Mike-EEE** [comment]
It's close!  My concern here is that it looks like it's being destructured?  It's hard to see how the item will be rendered, unless it is known that the property order is the order in which objects are rendered.

Also, I have begun lately to use `Project` rather than `Transform` ... thanks to some nice code/guidance that I've spied upon by @nblumhardt lately. :D  

I also do like `Format` as it seems stronger to my scenarios, but I am being selfish and not considering others in this regard!

**Comment by adamchester** [comment]
@Mike-EEE - so in your scenario, it might look more like:

```csharp
Log.Logger = new LoggerConfiguration()
    .Transform.ByType<MethodBase>(
        predicate: typeof(MethodBase).IsAssignableFrom,
        transformation: m => $"{m.DeclaringType.Name}.{m.Name}")
    .WriteTo.Sink(new MySink())
    .CreateLogger();
```

You have chosen to convert one scalar to another scalar early in the pipeline. From there onwards, all sinks and formatters have no choice but to respect your wishes, right?

Note I'm not sure about the naming of `.Transform.ByType<T>` - just trying to make a point at this stage.

**Comment by Mike-EEE** [comment]
Exactly!  Basically an application-specific override that is checked first, otherwise, behave as normal.  Much like. :smile: 

Yeah, I used `Transform` for the longest time in my framework (and honestly I am a little burned out on the word LOL), but now I use a much more terse `With` command, but `Project` is exactly what it is. :)

**Comment by adamchester** [comment]
@Mike-EEE did you see #728 ?

**Comment by Mike-EEE** [comment]
Very cool!  Yes, that works, too. :smile:

**Comment by nblumhardt** [comment]
+1 on the PR. I think the two separate issues are still interesting to track:

 - [ ] Reflection types don't always fit "scalar" definition, should be transformable when `@Destructuring`
 - [ ] Text-based sinks would benefit from some kind of additional formatting mechanism (this issue, I think)

The first is a, straightforward change; the second probably not so, even though the notion makes sense. If the PR unblocks you @Mike-EEE I think the second item will probably remain unaddressed for now, as it's non-trivial to put a coherent design together for it - or at least seems that way first thing on a Sunday morning :-)

**Comment by merbla** [comment]
Watching from the stands on this one. @Mike-EEE did you care "when" in the pipeline these operations occurred or just that is was respected by the end of the pipeline?

**Comment by adamchester** [comment]
Regarding text-based formatting, I think it needs some more example scenarios. For some reason I thought [`ICustomFormatter`](https://msdn.microsoft.com/en-us/library/system.icustomformatter(v=vs.110).aspx) was going to help, when [I was looking at MessageTemplates](https://github.com/adamchester/MessageTemplates/blob/master/FsMessageTemplates/MessageTemplates.fs#L636).

**Comment by Mike-EEE** [comment]
Hi @merbla thanks for joining in. :smile: Thinking about this I am thinking that whenever a token is rendered -- no matter where or when -- it first looks to the "overrides" and if present uses those.  But, that's just my take!

@nblumhardt This looks good to me, but of course I have to try it to be 100% certain. :D  Maybe you can help this newb with some guidance.  I spoke with @adamchester briefly on Gitter, but is there an easy/recommended/assumed way to easily try new PR's without having to reference a local build/path on disk?  That is, currently I have my Serilog reference pointing to a NuGet package.  In order to try the PR as I understand it, I have to pull the code down, build it locally, and then update my local project to point to this new build.  Is there a more streamlined way of doing this from a NuGet perspective?  It would be more ideal to simply point to a PR-version of the package and try it out.

Any links/references/guidance on how to do this would be greatly appreciated!

**Comment by adamchester** [comment]
@Mike-EEE appveyor is where serilog builds happen, including PRs like #728.

have you seen this? https://www.appveyor.com/docs/nuget#project-nuget-feeds

**Comment by Mike-EEE** [comment]
WOW!  Very cool!!!  OK, that is totally the piece I was missing from this.  That's awesome!  I will give this a shot when I can and let you know if I have any questions/problems.  Thanks for the heads up @adamchester!

**Comment by nblumhardt** [comment]
I think #728 covered the initial part of this, and #730 (new issue) covers the rest, so closing this one. Feel free to jump in if I missed some part of it that needs addressing otherwise.

**Comment by Mike-EEE** [comment]
Hey everyone.  I wanted to let you know that I **FINALLY** just now got a chance to work with this, and it works **perfectly** as expected!  I can choose to use a `@` in front of the method parameter in my template, to determine if I want to emit it as a scalar or not.  This is very duh/obvious, but I was not anticipating this option/flexibility.  Super super awesome.

Great job, team!  Thank you so much (once again!) @adamchester & @nblumhardt !
